tests/drivers/pwm/pwm_loopback/src/test_pwm_loopback.c

```diff
@@ -51,14 +51,14 @@ void test_capture(uint32_t period, uint32_t pulse, enum test_pwm_unit unit,
 		TC_PRINT("Testing PWM capture @ %u/%u nsec\n",
 			 pulse, period);
 		err = pwm_pin_set_nsec(out.dev, out.pwm, period,
-				       pulse, out.flags);
+				       pulse, out.flags | PWM_FORCE_UPDATE);
```

Comment:
Is it possible that some drivers perform an immediate update? Does PWM specify the behavior? cc: @henrikbrixandersen

Reply:
Yes. As also noted in the description of this PR, this issue was discovered while implementing PWM capture on STM32 due to failing PWM loopback tests. 

The reason these tests fails on STM32 and not on e.g. NXP boards, is that whenever the PWM generator period is changed on NXP, the change takes affect immediately (as opposed to simple pulse width changes, where the PWM generation is merely updated in the next pulse). 

I am not sure introducing a new flag for handling this situation is needed. Instead, I would suggest trying to pursue the same path for STM32 as is taken by the NXP drivers; if the period is changed, the change is applied immediately.

Reply:
@henrikbrixandersen,
Having change taking effect immediately when period is changed may truncate the ongoing pulse and generate an undetermined pulse length. This is something generally not advice to have undetermined pulse. Even for a simple led blinking it will generate transient flash.
This is the reason why on STM32 by default changes on PWM are synchronous with timer rollover. This can still be changed but from my point of view it must be done on purpose. So a flag to force immediate update sounds good to me.

Reply:
I also think that synchronous behavior is preferred by default. In any case, the API should likely define the behavior.

Reply:
Thanks for the feedback!
As the expectations on the handling seem to differ and supplying the additional flag might lead to some extra effort by the user.
An alternative could be a dt-property that selects "always synchronous update"/"asynchronous update on period change", and can optionally be implemented by device drivers.
Would this be favored over the introduction of an additional flag?

Reply:
> Would this be favored over the introduction of an additional flag?

DT is supposed to provide information on hardware description or boot time configuration. In this case, we're talking about driver implementation, so I don't think this would be appropriate.

Reply:
I agree with @erwango that if we introduce an API for this, it should be through the PWM flags.

I think, however, that we need more discussion on how to engineer this API. It would seem that we currently have two different behaviors in Zephyr PWM controller drivers when changing the PWM period:
- Some drivers/controllers change the PWM period immediately.
- Some drivers/controllers change the PWM period on the next timer overflow (or similar).

While I do agree in general that the second option seems to be the best option, there are a few use-cases to keep in mind:
- Controllers with multiple PWM channels but only one timer.
- Controllers without support for synchronously changing the period.

Also worth noting is that (at least in my use-cases and understanding), when using PWM in regulation, LEDs, etc. the period is usually fixed and only the pulse width is changed. I believe all existing drivers update the pulse width synchronously.

Reply:
For reference, discussions we had when STM32 implemented synchronous behavior:

https://github.com/zephyrproject-rtos/zephyr/pull/29619
https://github.com/zephyrproject-rtos/zephyr/pull/29470